libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs

```diff
@@ -168,3 +181,202 @@ main :: IO ()
+  [GovActionState era]
+useTriples pairs as gs = zipWith3 help pairs as gs
+  where
+    help (parent, idx) a g =
+      g
+        { gasId = idx
+        , gasChildren = (children idx pairs)
+        , gasAction = setActionId a parent
+        }
+
+-- | [Pred era] that generate a valid Set(GovActionState era)
+govStatePreds :: forall era. (ConwayEraPParams era, Reflect era) => Proof era -> [Pred era]
+govStatePreds p =
+  [ MetaSize (SzRng 2 5) numActions
+  , Sized numActions gaids
+  , Subset gaids govActionIdUniv
+  , GenFrom pairs (Constr "genTree" (genTree . Set.toList) ^$ gaids)
+  , ListWhere
+      numActions
+      preGovstates
+      govActionStateTarget
+      [ Member (Left idV) govActionIdUniv
+      , Subset (Dom committeeVotesV) hotCommitteeCredsUniv
+      , Sized (Range 0 3) (Dom committeeVotesV)
+      , Subset (Dom drepVotesV) voteUniv
+      , Sized (Range 0 3) (Dom drepVotesV)
+      , Subset (Dom stakePoolVotesV) poolHashUniv
+      , Sized (Range 0 3) (Dom stakePoolVotesV)
+      , proposalDeposit p :=: depositV
+      , Random returnAddrV
+      , Oneof
+          actionV
+          [ (1, noConfidenceT, [Random gaPrevId])
+          ,
+            ( 1
+            , updateCommitteeT
+            ,
+              [ Random gaPrevId
+              , Sized (Range 0 3) gaRemMember
+              , Subset gaRemMember coldCommitteeCredsUniv
+              , Sized (Range 0 3) (Dom gaAddMember)
+              , Subset (Dom gaAddMember) coldCommitteeCredsUniv
+              , Random gaThreshold
+              ]
+            )
+          ]
+      , currentEpoch :=: proposedInV
+      , Random expiresAfterV
+      , Sized (Range 0 3) childrenV
+      , Subset childrenV govActionIdUniv
+      ]
+  , ListWhere
+      numActions
+      govActions
+      (Invert "GovAction" (typeRep @(GovAction era)) id :$ Partial govAction Just)
+      [ Oneof
+          govAction
+          [ (1, noConfidenceT, [Random gaPrevId])
+          ,
+            ( 1
+            , updateCommitteeT
+            ,
+              [ Random gaPrevId
+              , Sized (Range 0 3) gaRemMember
+              , Subset gaRemMember coldCommitteeCredsUniv
+              , Sized (Range 0 3) (Dom gaAddMember)
+              , Subset (Dom gaAddMember) coldCommitteeCredsUniv
+              , Random gaThreshold
+              ]
+            )
+          ]
+      ]
+  , govstates :<-: (Constr "useTriples" useTriples ^$ pairs ^$ govActions ^$ preGovstates)
+  ]
+  where
+    gaids = Var (pV p "gaids" (SetR GovActionIdR) No)
+    pairs = Var (pV p "pairs" (ListR (PairR (MaybeR GovActionIdR) GovActionIdR)) No)
+    numActions = Var (pV p "numActions" SizeR No)
+    preGovstates = Var (V "preGovstates" (ListR GovActionStateR) No)
+    govActions = Var (pV p "govActions" (ListR (GovActionR)) No)
+    govAction = Var (pV p "govAction" GovActionR No)
+
+fstL :: Lens' (a, b) a
+fstL = lens fst (\(_, b) a -> (a, b))
+
+sndL :: Lens' (a, b) b
+sndL = lens snd (\(a, _) b -> (a, b))
```

Comment:
Don't we already have those lenses in `Constrained.Vars`?

Reply:
These have been moved to the Lenses.hs file  (comments and all)